google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -0,0 +1,199 @@
+    gaeAppFlex("gae_app_flex"),
+    gaeAppStandard("gae_app_standard"),
+    gceInstance("gce_instance"),
+    global("global");
+
+    private final String key;
+
+    Resource(String key) {
+      this.key = key;
+    }
+
+    String getKey() {
+      return key;
+    }
+  }
+
+  private static Map<String, Label[]> resourceTypeWithLabels;
+
+  static {
+    resourceTypeWithLabels =
+        new ImmutableMap.Builder<String, Label[]>()
+            .put(
+                Resource.gaeAppFlex.getKey(),
+                new Label[] {
+                  Label.instanceName,
+                  Label.moduleId,
+                  Label.versionId,
+                  Label.instanceName,
+                  Label.zone
+                })
+            .put(
+                Resource.gaeAppStandard.getKey(),
+                new Label[] {Label.appId, Label.moduleId, Label.versionId})
+            .put(Resource.container.getKey(), new Label[] {Label.clusterName, Label.zone})
+            .put(Resource.gceInstance.getKey(), new Label[] {Label.instanceId, Label.zone})
+            .build();
+  }
+
+  /* Return a self-configured monitored Resource. */
+  static MonitoredResource getResource(String projectId, String resourceTypeParam) {
+    String resourceType = resourceTypeParam;
+    if (Strings.isNullOrEmpty(resourceType)) {
+      Resource detectedResourceType = getAutoDetectedResourceType();
+      resourceType = detectedResourceType.getKey();
+    }
+    // Currently, "gae_app" is the supported logging Resource type, but we distinguish
+    // between "gae_app_flex", "gae_app_standard" to support zone id, instance name logging on flex VMs.
+    // Hence, "gae_app_flex", "gae_app_standard" are trimmed to "gae_app"
+    String resourceName = resourceType.startsWith("gae_app") ? "gae_app" : resourceType;
+    MonitoredResource.Builder builder =
+        MonitoredResource.newBuilder(resourceName).addLabel(Label.projectId.getKey(), projectId);
+    Label[] resourceLabels = resourceTypeWithLabels.get(resourceType);
+    if (resourceLabels != null) {
+      for (Label label : resourceLabels) {
+        String value = getValue(label);
+        if (value != null) {
+          builder.addLabel(label.getKey(), value);
+        }
+      }
+    }
+    return builder.build();
+  }
+
+  /* Detect monitored Resource type using environment variables, else return global as default. */
+  private static Resource getAutoDetectedResourceType() {
+    if (System.getenv("GAE_INSTANCE") != null) {
+      return Resource.gaeAppFlex;
+    }
+    if (System.getenv("KUBERNETES_SERVICE_HOST") != null) {
+      return Resource.container;
+    }
+    if (ServiceOptions.getAppEngineAppId() != null) {
+      return Resource.gaeAppStandard;
+    }
+    if (MetadataConfig.getInstanceId() != null) {
+      return Resource.gceInstance;
+    }
+    // default Resource type
+    return Resource.global;
+  }
+
+  protected static List<LoggingEnhancer> getResourceEnhancers() {
```

Comment:
This can just be package private - `protected` doesn't add anything for a static method.